### PR TITLE
[JENKINS-65117] fix the case sensitivity in the test method for the DN

### PR DESCRIPTION
**Pull request by jtnord**
[JENKINS-65117](https://issues.jenkins.io/browse/JENKINS-65117)

A DN is hierachical and contains attribute names and their values.  

An attribute name is case insensitive so when comparing a String representaion of a DN you will have incorrect results if one side has a different case for the attribute name than the other.

The solution is to parse the DN and compare the parsed results.

Manually tested with the planet express test DB and with a user that has some characters that need to be escaped as part of its CN (a comma).

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/jenkins/security/plugins/ldap/LDAPExtendedTemplate.java [hunk 1]
```diff
@@ -60,7 +60,7 @@ public LDAPExtendedTemplate(ContextSource dirContextFactory) {
      * @return the first matching entry converted using the specified {@link LdapEntryMapper}, or null if no matching
      * entry was found.
      */
-    public <T> @CheckForNull T searchForFirstEntry(@NonNull final String base, @NonNull final String filter,
+    public @CheckForNull <T> T searchForFirstEntry(@NonNull final String base, @NonNull final String filter,
```
Comment (jtnord) [inline src/main/java/jenkins/security/plugins/ldap/LDAPExtendedTemplate.java:63]: eclipse would not compile without this..
```
Unresolved compilation problem:
        The annotation @CheckForNull is disallowed for this location
```

File: src/main/resources/jenkins/security/plugins/ldap/Messages.properties [hunk 2]
```diff
@@ -7,7 +7,7 @@ LDAPSecurityRealm.AuthenticationSuccessful=Authentication: successful
 LDAPSecurityRealm.AuthenticationFailed=Authentication: failed for user "{0}"
 LDAPSecurityRealm.AuthenticationFailedEmptyPass=Authentication: failed for user "{0}" with empty password
 LDAPSecurityRealm.UserId=User ID: {0}
-LDAPSecurityRealm.UserDn=User Dn: {0}
+LDAPSecurityRealm.UserDn=User DN: {0}
```
Comment (jtnord) [inline src/main/resources/jenkins/security/plugins/ldap/Messages.properties:10]: always `DN` never `Dn`

**Review by car-roll — APPROVED** [review]

File: src/main/java/hudson/security/LDAPSecurityRealm.java [hunk 3]
```diff
@@ -1778,11 +1779,25 @@ public FormValidation validate(LDAPSecurityRealm realm, String user, String pass
...
-                                    loginDetails.getDn(), lookUpDetails.getDn()));
-                    potentialLockout = true; // consistency is important
+                try {
+                    // do not use string comparison here as Attributes are always case insentive
+                    // the values may or may not be and that depend in part on the attributetype
+                    // and configuration
+                    LdapName loginDN = new LdapName(loginDetails.getDn());
+                    LdapName lookupDN = new LdapName(lookUpDetails.getDn());
+                    if (!loginDN.equals(lookupDN)) {
+                        // do not use the loginDN.toString() directory as this returns the unparsed version
```
Comment (jtnord) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1789]: I guess I should address this note to self

**Review by kerogers-cloudbees — APPROVED** [review]
Change looks good to me.

File: src/main/java/hudson/security/LDAPSecurityRealm.java (hunk 3 above)
Comment (jtnord) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1789]:
```suggestion
                        // TODO do not use the loginDN.toString() directory as this returns the unparsed version
```

**Review by MRamonLeon — APPROVED** [review]

**Review by bitwiseman — APPROVED** [review]

**Review by rsandell — APPROVED** [review]

**Pull request merged**